### DISCUSSION TITLE
Batch import BMP/Alpha and export as Targa with Alpha

### DISCUSSION THREAD

**Discussion by danutzzzzz** [ImageMagick/ImageMagick#7682] [Help] [answered]
### ImageMagick version

7.1.1

### Operating system, version and so on

Windows 11 [Version 10.0.22631.4112]

### Description

I have a large selection of .tex container files that contain .bmp images, often with alpha masks that i have to edit and then recompile and export into a TGA format and then re-add to .tex containers.

I have the .tex extraction and recompiling automated, its the stage with DXTBmp that's currently manual that i wish to move over too and use/automate with ImageMagick.

**Files:** https://github.com/daNutzzzzz/Textures

**Example1** 
1. To extract the .bmp files from the .tex container, you run: `text2pic /f BANNERLCstat.tex`

2. BMP files extracted are:
BANNERLCstat.bmp
BANNERLCstat@.bmp (Alpha)

3. Import both .BMP into DXTBmp and save as Targa with Alpha (TGA).

4. Add TGA file back into .tex container:
`pic2tex -key -mipmap9 BANNERLCstat.tex`

**Example2**
Same as Example1 but to recompile:

1. Add TGA files to individual .tex containers:
```
for %%i in (*.tga) do (
    pic2tex -key -mipmap9 %%i
)
```

2. Add all single .tex files to one .tex container:
`copy /b .\mipmapheader.tex+0*.tex afterburn.tex`

**Comment by snibgo** [comment]
> ... its the stage with DXTBmp that's currently manual that i wish to move over too and use/automate with ImageMagick.

Is there a description somewhere of exactly what processing DXTBmp does in your use case? IM can, of course, read BMP files and write TGA.

**Reply by danutzzzzz** [reply]
> > ... its the stage with DXTBmp that's currently manual that i wish to move over too and use/automate with ImageMagick.
> 
> Is there a description somewhere of exactly what processing DXTBmp does in your use case? IM can, of course, read BMP files and write TGA.

Hi, i assume its simply combining the two files, removing some "junk" in the alpha file then saving as TGA?

Ive uploaded the TGA file in this example: https://github.com/daNutzzzzz/Textures/blob/main/Example1/BANNERLCstat/BANNERLCstat.tga

**Comment by snibgo** [comment]
Thanks for the link.

Looking at the two BMP files, and the TGA file, it seems that one BMP is used as the alpha for the other BMP. So:

```
magick BANNERLCstat.bmp BANNERLCstat@.bmp -alpha off -compose CopyOpacity -composite out.tga
```
To check the result, we can compare the result out.tga with the supplied tga:
```
magick compare -metric RMSE BANNERLCstat.tga out.tga NULL:
0 (0)
```
The result image is identical to the expected result.

**Reply by danutzzzzz** [reply]
> Thanks for the link.
> 
> Looking at the two BMP files, and the TGA file, it seems that one BMP is used as the alpha for the other BMP. So:
> 
> ```
> magick BANNERLCstat.bmp BANNERLCstat@.bmp -alpha off -compose CopyOpacity -composite out.tga
> ```
> 
> To check the result, we can compare the result out.tga with the supplied tga:
> 
> ```
> magick compare -metric RMSE BANNERLCstat.tga out.tga NULL:
> 0 (0)
> ```
> 
> The result image is identical to the expected result.

Hello, thank you for this, ive finally had time to check and it is "working" by comparing visually, however i need to set Origin as "Bottom - Left".. currently the image being produced is set to "Top - Left"

If you check the binary of two example files [here](https://github.com/daNutzzzzz/Textures/tree/main/Example%203) you will see the imagemagick is reversed.

How do i set this in the command line?

**Reply by fmw42** [reply]
Add `-gravity southwest` before -compose

see https://imagemagick.org/script/command-line-options.php#gravity

**Reply by danutzzzzz** [reply]
Thanks but that hasnt done anything... reading the article it seems to refer to overlaying position of text or an image?

**Reply by fmw42** [reply]
Not just text but also images.  Try also adding +repage before -alpha off.  I am not sure if BMP supports a virtual canvas.

What are the sizes of the two BMP images? Can you post them?  Zip them if necessary to upload.

**Reply by danutzzzzz** [reply]
The images are here: https://github.com/daNutzzzzz/Textures/tree/main/Example%203

**Reply by fmw42** [reply]
Sorry, but I could not open either BMP on my Mac OSX Ventura.  Imagemagick says improper header. I also tried several other tools including Photoshop and none could open them. They all report invalid BMP files.

**Reply by danutzzzzz** [reply]
Strange, windows native apps opens them fine, so does GIMP.. so do any of these [online tools](https://www.bing.com/search?pglt=169&q=bmp+viewer+online&cvid=10924340e88f4901b96848f7b5f3e92b&gs_lcrp=EgZjaHJvbWUyBggAEEUYOdIBCDM0NzBqMGoxqAIAsAIA&FORM=ANNTA1&PC=U531)


You can even view them on GitHub:
https://raw.githubusercontent.com/daNutzzzzz/Textures/refs/heads/main/Example%203/00000.bmp
https://raw.githubusercontent.com/daNutzzzzz/Textures/refs/heads/main/Example%203/00000@.bmp

**Reply by danutzzzzz** [reply, marked as answer]
Ok so ive managed to fix most of it by using `-flip` instead of `-gravity`... however the file produced has one byte thats incorrect and stopping it from importing:

![image](https://github.com/user-attachments/assets/b60ae521-dd73-4928-9b4c-28d42e17cc75)

Any idea what this byte is setting? Answer: https://github.com/ImageMagick/ImageMagick/discussions/7682#discussioncomment-10930785

**Reply by fmw42** [reply]
What have you posted.  That is not a valid image and that is what I see when I try to open it in several tools.

I can post any normal BMP. But something is odd about your files.

-flip should not cause any issues if you have a valid BMP file.  I am not sure your TGA to BMP processing is valid.

**Reply by fmw42** [reply]
Also your BANNERLCstat@.bmp link downloads as a mail address due to the @ symbol.  Why do you use that?  Perhaps change the symbol.

**Reply by danutzzzzz** [reply]
Well its all working perfectly fine my side and for thousands of others.

**Reply by fmw42** [reply]
I cannot even open your BANNERLCstat.tga file.  Same issue invalid file format.

**Reply by fmw42** [reply]
Is it working because you are using images before uploading?  mIght the issue be the upload process?

**Reply by fmw42** [reply]
Why don't you try zipping those files and uploading rather than uploading as you have unzipped.

**Reply by danutzzzzz** [reply]
Never known github to modify any file uploaded.

Heres the zip: https://github.com/daNutzzzzz/Textures/raw/refs/heads/main/Example%203/Example%203.zip

**Reply by fmw42** [reply]
OK. I got them from your GITHUB posts and those open.

**Reply by fmw42** [reply]
I processed them just as snibgo did as both tga and bmp.

`magick 00000.bmp 00000@.bmp -alpha off -compose copy_opacity -composite result.bmp`

`magick 00000.bmp 00000@.bmp -alpha off -compose copy_opacity -composite result.tga`

Here are the results.  Do they look correct? The details in the alpha file look correct to me to match the locations in the image file.  Is that what you need?

[Archive.zip](https://github.com/user-attachments/files/17356873/Archive.zip)

**Reply by fmw42** [reply]
Looks to me that they were not properly uploaded, except when posting them directly to GIT

**Reply by fmw42** [reply]
The ones in your zip file load OK.  I think it is when you tried to post as folders that have issues.

**Reply by danutzzzzz** [reply]
> Looks to me that they were not properly uploaded, except when posting them directly to GIT

not sure what you mean, all files are posted directly to GitHub.

**Reply by danutzzzzz** [reply]
> I processed them just as snibgo did as both tga and bmp.
> 
> `magick 00000.bmp 00000@.bmp -alpha off -compose copy_opacity -composite result.bmp`
> 
> `magick 00000.bmp 00000@.bmp -alpha off -compose copy_opacity -composite result.tga`
> 
> Here are the results. Do they look correct? The details in the alpha file look correct to me to match the locations in the image file. Is that what you need?
> 
> [Archive.zip](https://github.com/user-attachments/files/17356873/Archive.zip)

No these are the same as  i mention here: https://github.com/ImageMagick/ImageMagick/discussions/7682#discussioncomment-10929544

**Reply by danutzzzzz** [reply]
> Ok so ive managed to fix most of it by using `-flip` instead of `-gravity`... however the file produced has one byte thats incorrect and stopping it from importing:
> 
> ![image](https://private-user-images.githubusercontent.com/89404081/376063435-b60ae521-dd73-4928-9b4c-28d42e17cc75.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg4NTQ1MTMsIm5iZiI6MTcyODg1NDIxMywicGF0aCI6Ii84OTQwNDA4MS8zNzYwNjM0MzUtYjYwYWU1MjEtZGQ3My00OTI4LTliNGMtMjhkNDJlMTdjYzc1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDEzVDIxMTY1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcxYTEzODdlYzRjZGYxNTBmMTAzYjI5YjMwMDExMzEyMDc3ZWNhZGM3YjA5NGVkNzE5NzFiOWM3NWE0NDZiZTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.OpvFJ6noI7Tp2U4zIpsQUOl6oAmsBLgHcqh7DDZSIck)
> 
> Any idea what this byte is setting?

Ive figured out this byte represents the origin location, 28 is Top Right and 08 is Bottom Left, ive written a script to correct it.

**Comment by danutzzzzz** [comment]
Just want to say thank you @snibgo you helped me figure this out.

**Comment by snibgo** [comment]
> Ive figured out this byte represents the origin location, 28 is Top Right and 08 is Bottom Left

You can set orientation in IM with "-orient bottom-left". See https://imagemagick.org/script/command-line-options.php#orient

**Reply by danutzzzzz** [reply]
This worked perfectly! thank you again, i can now remove my custom script from the process.
